src/model/validation.rs

```diff
@@ -0,0 +1,424 @@
+                HashMap::from([
+                    (
+                        "bus".into(),
+                        HashMap::from([("lattice".into(), ()), ("host".into(), ())]),
+                    ),
+                    (
+                        "messaging".into(),
+                        HashMap::from([("consumer".into(), ()), ("handler".into(), ())]),
+                    ),
+                ]),
+            ),
+        ])
+    })
+}
+
+static MANIFEST_NAME_REGEX_STR: &str = r"^[-\w]+$";
+static MANIFEST_NAME_REGEX: OnceLock<Regex> = OnceLock::new();
+
+/// Retrieve regular expression which manifest names must match, compiled to a usable [`Regex`]
+fn get_manifest_name_regex() -> &'static Regex {
+    MANIFEST_NAME_REGEX.get_or_init(|| {
+        Regex::new(MANIFEST_NAME_REGEX_STR)
+            .context("failed to parse manifest name regex")
+            .unwrap()
+    })
+}
+
+/// Check whether a manifest name matches requirements, returning all validation errors
+pub fn validate_manifest_name(name: &str) -> impl ValidationOutput {
+    let mut errors = Vec::new();
+    if !get_manifest_name_regex().is_match(name) {
+        errors.push(ValidationFailure::new(
+            ValidationFailureLevel::Error,
+            format!("manifest name [{name}] is not allowed (shoudl match regex [{MANIFEST_NAME_REGEX_STR}])"),
+        ))
+    }
+    errors
+}
+
+/// Check whether a manifest name matches requirements
+pub fn is_valid_manifest_name(name: &str) -> bool {
+    validate_manifest_name(name).valid()
+}
+
+/// Check whether a manifest version is valid, returning all validation errors
+pub fn validate_manifest_version(version: &str) -> impl ValidationOutput {
+    let mut errors = Vec::new();
+    if version == LATEST_VERSION {
+        errors.push(ValidationFailure::new(
+            ValidationFailureLevel::Error,
+            format!("{LATEST_VERSION} is not allowed in wadm"),
+        ))
+    }
+    errors
+}
+
+/// Check whether a manifest version is valid requirements
+pub fn is_valid_manifest_version(version: &str) -> bool {
+    validate_manifest_version(version).valid()
+}
+
+/// Check whether a known grouping of namespace, package and interface are valid.
+/// A grouping must be both known/expected and invalid to fail this test (ex. a typo).
+///
+/// NOTE: what is considered a valid interface known to the host depends explicitly on
+/// the wasmCloud host and wasmCloud project goals/implementation. This information is
+/// subject to change.
+fn is_invalid_known_interface(namespace: &str, package: &str, interface: &str) -> bool {
+    let known_interfaces = get_known_interface_lookup();
+    let Some(pkg_lookup) = known_interfaces.get(namespace) else {
+        // This namespace isn't known, so it may be a custom interface
+        return false;
+    };
+    let Some(iface_lookup) = pkg_lookup.get(package) else {
+        // Unknown package inside a known interface we control is probably a bug
+        return true;
+    };
```

Comment:
I'm a _little_ concerned with this one when it comes to wadm supporting interfaces in a forwards compatible way, since this would also necessitate a wadm upgrade when, say, `wasi:keyvalue` adds a new interface 🤔 

I totally get what this is supposed to prevent though, linking on like `wasi:http/incoming-handlerr` and being frustrated endlessly. Do you have thoughts on the tradeoff here?

Reply:
What I did for now was to make it a warning -- it can be up to downstream tools on whether to be strict or not

Reply:
What I did for now was to make it a warning -- it can be up to downstream tools on whether to be strict or not, what do you think @brooksmtownsend

Reply:
@vados-cosmonic I think that makes sense! So here a future inclusion of `wasi:nosql` for example would just be a "heads up, we didn't know about this one", but adding `wasi:keyvalue/eventual` (the old interface) would fail validation and catch it. LGTM